### PR TITLE
[Transformations] Complete GQA windowed-cache support: staging + attention_bias

### PR DESCRIPTION
**Pull request by ankitm3k**
## Details

Follow-up to #37207 (GroupQueryAttention sliding-window support). It closes the two remaining gaps in the
windowed KV-cache path of `GroupQueryAttentionDecomposition`: the multi-token **staging** regime, and
**attention_bias combined with a windowed cache**. Two commits.

## Commit 1 — multi-token sliding_window_cache staging

Before this change, the windowed branch (`sliding_window_cache=1`) only handled steps that fit the window:
it built `present = [survivors, new, zeros]` in place with `kept = end_after - S`. A multi-token step that
overflows the window in one shot (prefill into a window-sized cache) makes `kept` negative, so the in-place
indices are invalid. That case (the ONNX Runtime "staging" regime) was rejected for a statically-known `S>1`
and silently mis-handled on the dynamic-shape path.

## What changed

The windowed branch now dispatches on the query sequence length:

- **Static single-token decode (`S == 1`)** keeps the existing in-place `Gather` + `ScatterUpdate` assembly
  (static-shape friendly), unchanged.
- **Otherwise (dynamic `S`)** attention runs against a temporary buffer of the `end_before` resident rows
  plus the `S` new tokens (`Slice` + `Concat`), and only the surviving tail (the last `end_after` rows) is
  written back into the capacity-C present cache. This matches the ONNX Runtime CPU kernel's staging path
  (`PlanWindowedKvCache` in `contrib_ops/cpu/bert/group_query_attention.cc`) and is correct for any step,
  including one that crosses a window eviction.

The attention mask is reused unchanged — causal and local-window masking depend only on the query-key
distance, so the cache-relative coordinate system (`mask_past_seqlen = end_before`) is exact. The returned
present cache stays capacity-C; the staging buffer is internal.

A statically-known `S>1` with `sliding_window_cache` stays rejected (the runtime-sized temp buffer is not
static-shape lowerable, e.g. for NPU); the dynamic path was already permitted and is now correct.

## Commit 2 — attention_bias with sliding_window_cache

`attention_bias` was rejected together with `sliding_window_cache` because the bias is indexed by absolute
key position, while a windowed (rolling) cache's first slot is not absolute key 0 after an eviction, so
slicing the bias from column 0 misaligned it.

Align the bias to the cache instead of rejecting: `make_attention_mask` now slices the bias starting at
`bias_col_offset = P - resident_rows` (the absolute key of the KV buffer's first slot) — `0` for a
full-length cache (unchanged), `P - kept` for the windowed decode path, and `P - end_before` for the staging
path. The frontend no longer rejects the combination.

## Testing

Added ONNX FE numerical tests verified against a NumPy port of the ORT windowed attention:
- `onnx_model_gqa_sliding_window_cache_staging` — float, dynamic `S=6`, capacity 4, window 2 (crosses an
  eviction; the in-place `kept` would be negative).
- `onnx_model_gqa_sliding_window_cache_staging_i8` — int8 quantized KV, same step; checks the quantize
  round-trip is preserved through staging (present bytes exact).
- `onnx_model_gqa_sliding_window_cache_bias` — windowed decode with a nonzero bias column offset (P=5).
- `onnx_model_gqa_sliding_window_cache_bias_staging` — bias on the staging path (nonzero offset).

All pass on **CPU** and **INTERPRETER**. Additionally verified off-tree (numpy oracle) across i4/f8
quantized KV, rotary (`do_rotary=1`), boundary/multi-cycle cases (`x==C`, first eviction, `S=10` multi-cycle,
`P>0` resident-past), and bias at nonzero offsets on both decode and staging — all numerically exact. The
now-obsolete negative test that asserted `attention_bias + sliding_window_cache` throws is removed.

**GPU:** these dynamic-shape windowed tests are skipped. On GPU the windowed present cache comes back zeroed
(attention output is correct). The isolated present-write subgraph (`windowed_cache_end` + `Range` + `Slice`
+ `ScatterUpdate`) lowers correctly on GPU on its own; the failure only appears in the full assembled dynamic
graph. This is a GPU-plugin dynamic-shape issue, not a decomposition bug (CPU/INTERPRETER are numerically
exact) — tracked as a GPU follow-up. The static-shape single-token windowed path is unaffected and runs on GPU.

## Tickets
- CVS-192121 (follow-up)

### AI Assistance:
Yes — used for reading the ONNX Runtime MLAS CPU / CUDA kernels and spec, mapping them onto OpenVINO
op primitives, generating reference tensors from ONNX Runtime, adversarial correctness review, and
the cross-device (CPU/GPU/NPU) verification.

### CODE REVIEW

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp [hunk 1]
```diff
@@ -167,53 +167,92 @@ ov::OutputVector ov::pass::GroupQueryAttentionDecomposition::decompose(
...
+                                                              one_without_shape,
+                                                              ov::element::i64);
+            const auto new_idx = register_new_node<v1::Add>(new_row, kept);
+
+            present_k = register_new_node<v3::ScatterUpdate>(zeros, kept_idx, survivor_k, scatter_axis);
+            present_k = register_new_node<v3::ScatterUpdate>(present_k, new_idx, K, scatter_axis);
+            present_v = register_new_node<v3::ScatterUpdate>(zeros, kept_idx, survivor_v, scatter_axis);
+            present_v = register_new_node<v3::ScatterUpdate>(present_v, new_idx, V, scatter_axis);
+
+            K = present_k;
```
Comment (bumbosiepsak) [inline src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:221]: In the static single-token branch, K becomes the capacity-C buffer, so kv_len is C:
```
// line 221-226 (static_single_token branch)
K = present_k;
V = present_v;
mask_past_seqlen = register_new_node<v0::Unsqueeze>(kept, zero);
// First resident slot holds absolute key P - kept (the survivors start there).
bias_col_offset =
    register_new_node<v0::Unsqueeze>(register_new_node<v1::Subtract>(abs_past_scalar, kept), zero);
```
```
// line 286 — kv_len is the capacity C here, because K = present_k (capacity-sized)
const auto concat_kv_len = get_dimensions(K.get_node_shared_ptr(), {2});
```

`make_attention_mask` then slices the bias `[bias_col_offset : bias_col_offset + `C]:
```
// line 459-460 (make_attention_mask, has_bias branch)
const auto bias_stop = register_new_node<v1::Add>(bias_col_offset, kv_len_1d);
bias = register_new_node<v8::Slice>(bias, bias_col_offset, bias_stop, one, two);
```

The bias tensor only has `total_sequence_length` (= `P+S`) columns. With `bias_col_offset = P - kept, only kept + S = end_after` columns remain, so after any eviction (`end_after < C`) the `Slice` clamps to width `end_after` while `mask` is `[curr, C]` — the subsequent `Add(mask, bias)` gets mismatched trailing dims (`C` vs `end_after`) and cannot broadcast. The staging branch avoids this because there `kv_len = end_before + S` matches the contiguous absolute-key span. This path is reachable from the frontend but exercised by none of the new tests — all four reshape `query` to `{1,-1,64}`, routing through staging instead.

Suggested direction: in the static single-token branch, align the bias to the `[survivors|new|zeros]` layout (slice to width `kept + S`, zero-pad to `C`), and add a static-shape single-token + bias test that crosses a window eviction.

File: src/frontends/onnx/tests/onnx_import_com_microsoft.in.cpp [hunk 2]
```diff
@@ -5818,6 +5805,451 @@ OPENVINO_TEST(${BACKEND_NAME}, onnx_model_gqa_sliding_window_cache) {
...
 
+// Windowed cache, multi-token step that overflows the window (the "staging" regime): C=4, W=2, prefill S=6,
+// so end_after=3 and the in-place kept=end_after-S=-3 is invalid. Dynamic query seqlen routes into the
+// staging branch (a static S>1 is rejected at import). Reference is a NumPy port of the ORT windowed attention.
+OPENVINO_TEST(${BACKEND_NAME}, onnx_model_gqa_sliding_window_cache_staging) {
+    // On GPU the present cache comes back zeroed (attention output is correct); verified numerically exact on
+    // CPU/INTERPRETER. The failure only reproduces in the full staging graph - the isolated present-write
+    // subgraph (windowed_cache_end + Range + Slice + ScatterUpdate) lowers correctly on GPU on its own - so it
+    // is a GPU-plugin issue in the assembled dynamic graph, not a decomposition bug. Tracked as a GPU follow-up.
+    if (s_device == ov::test::utils::DEVICE_GPU) {
```
Comment (bumbosiepsak) [inline src/frontends/onnx/tests/onnx_import_com_microsoft.in.cpp:5821]: All four new tests skip GPU with the same guard:
```
// e.g. onnx_model_gqa_sliding_window_cache_staging
if (s_device == ov::test::utils::DEVICE_GPU) {
    GTEST_SKIP() << "GPU zeroes the staging present cache in the full graph; verified correct on CPU/INTERPRETER.";
}
```

Since this decomposition is what runs on GPU, a real model executed on GPU gets a zeroed `present_key`/`present_value`. In autoregressive decode the present cache is fed back as the next step's past, so subsequent tokens read zeros and silently produce wrong output with no error surfaced. Additionally, dynamic single-token decode now routes to the staging branch:
```
// line 190-192 — dynamic S (incl. dynamic single-token decode) takes the staging path
const auto& q_ps = node->get_input_partial_shape(0);
const bool static_single_token = q_ps.rank().is_static() && q_ps.rank().get_length() == 4 &&
                                 q_ps[2].is_static() && q_ps[2].get_length() == 1;
```

so if the previous in-place path lowered correctly on GPU, this may be a GPU decode regression for `sliding_window_cache` models.

Suggested direction: gate the windowed/staging decomposition away from GPU (or keep the in-place path on GPU for single-token) until the GPU dynamic-shape issue is fixed, so GPU users get a correct result or a clear failure instead of silent corruption. Link the tracking ticket for the GPU present-cache bug.

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp (hunk 1 above)
Comment (ankitm3k) [reply src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:221]: Fixed — `make_attention_mask` now zero-pads the sliced bias back up to `kv_len` whenever it clamps narrower. This also affects the plain full-length-cache branch the same way (its buffer can exceed `total_sequence_length` too), so the fix covers both, not just the windowed one. The padded columns are always past the causal/window edge, so they're already masked to `lowest()` regardless of bias value — zero is safe. Added `onnx_model_gqa_sliding_window_cache_bias_static`, your exact repro (static sequence_length==1, P=5/capacity=4/window=2, crossing an eviction): reproduced the broadcast failure before the fix, now passes numerically identical to the dynamic-shape case on CPU/GPU/INTERPRETER. (523688e3d2)

File: src/frontends/onnx/tests/onnx_import_com_microsoft.in.cpp (hunk 2 above)
Comment (ankitm3k) [reply src/frontends/onnx/tests/onnx_import_com_microsoft.in.cpp:5821]: Confirmed as a real regression, not the same limitation restated: at runtime `S==1` on a dynamic-shape query, the old in-place path was bit-exact on GPU; the new dispatch routes that case into staging, which GPU zeroes. Can't just revert to in-place for dynamic shapes though — the old code also hard-crashes on a real window-crossing step (CPU: negative `ScatterUpdate` range; GPU: `CL_OUT_OF_RESOURCES`), which is exactly what staging fixes. The real fix is dispatching on runtime `S` instead of declared shape (an `If` split) — too large for this cycle, so parking it as a follow-up. Reworded the skip messages to state the true scope; the escape hatch still works today — static `sequence_length==1` stays on the in-place path and is bit-exact on GPU. (523688e3d2)

File: src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp (hunk 1 above)
Comment (bumbosiepsak) [reply src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:221]: Consider adding a helper pass that will report an error:
```cpp
// src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp, before line 777
manager.register_pass<ov::intel_gpu::RejectUnsupportedGQAWindowedCache>();
```
with contents:
```cpp
// src/plugins/intel_gpu/src/plugin/transformations/reject_unsupported_gqa_windowed_cache.hpp

#pragma once

#include "openvino/pass/matcher_pass.hpp"

namespace ov::intel_gpu {

// Fails fast on a GroupQueryAttention configuration the GPU plugin cannot lower correctly:
// sliding_window_cache with a non-single-token (dynamic) query sequence length. That config
// decomposes (in CommonOptimizations) to a dynamic-shape staging subgraph whose present KV cache
// comes back zeroed on GPU, silently corrupting subsequent autoregressive decode steps. Raising a
// clear error at compile time is preferable to returning wrong results; the static single-token
// windowed path lowers correctly and is left untouched. Tracked as GPU follow-up CVS-192121.
class RejectUnsupportedGQAWindowedCache : public ov::pass::MatcherPass {
public:
    OPENVINO_MATCHER_PASS_RTTI("ov::intel_gpu::RejectUnsupportedGQAWindowedCache");
    RejectUnsupportedGQAWindowedCache();
};

}  // namespace ov::intel_gpu
```
```cpp
// src/plugins/intel_gpu/src/plugin/transformations/reject_unsupported_gqa_windowed_cache.cpp

RejectUnsupportedGQAWindowedCache::RejectUnsupportedGQAWindowedCache() {
    auto gqa_m = ov::pass::pattern::wrap_type<ov::op::internal::GroupQueryAttention>();

    ov::matcher_pass_callback callback = [](ov::pass::pattern::Matcher& m) {
        auto gqa = ov::as_type_ptr<ov::op::internal::GroupQueryAttention>(m.get_match_root());
        if (!gqa || !gqa->get_sliding_window_cache())
            return false;

        // Same predicate the decomposition uses: only a statically-known single-token decode
        // (Q layout [B, N, S, H], S == 1) takes the GPU-safe in-place branch. Anything else
        // (dynamic S, or S > 1) becomes the dynamic staging subgraph that GPU miscompiles.
        const auto& q_ps = gqa->get_input_partial_shape(0);
        const bool static_single_token = q_ps.rank().is_static() && q_ps.rank().get_length() == 4 &&
                                         q_ps[2].is_static() && q_ps[2].get_length() == 1;
        OPENVINO_ASSERT(static_single_token,
                        "GroupQueryAttention: sliding_window_cache=1 with a dynamic/multi-token query "
                        "sequence length is not supported on GPU (present KV cache is not lowered "
                        "correctly). Use CPU for this configuration; tracked as GPU follow-up CVS-192121.");
        return false;  // validation-only; never mutates the graph
    };

    auto matcher = std::make_shared<ov::pass::pattern::Matcher>(gqa_m, "RejectUnsupportedGQAWindowedCache");
    register_matcher(matcher, callback);
}
```

Test generation suggested too.

Comment (ankitm3k) [reply src/common/transformations/src/transformations/op_conversions/group_query_attention_decomposition.cpp:221]: I agree with the direction you propose. This lives in the GPU plugin (different codeowners/CI path than this PR's transformations/ONNX-FE scope), so I'll implement it as a small, fast-turnaround follow-up PR rather than fold it into this one — will link back here for context and tag you. Will include a regression test per your ask.

**Review by bumbosiepsak — APPROVED** [review]

**Comment by bumbosiepsak** [comment]
Agreed to deliver the accuracy guard (https://github.com/openvinotoolkit/openvino/pull/37281#discussion_r3751030361) in a follow-up delivery.
